thinc/layers/chain.py

```diff
@@ -18,13 +17,15 @@ def chain_no_types(*layer: Model) -> Model:
 
 
 def chain(
```

Comment:
I was a bit unsure here, so would be glad of careful reviews.

Reply:
Hmm to be perfectly honest I'm a little unsure myself here. The `XY_YZ_OutT` stuff was added in https://github.com/explosion/thinc/pull/255 and really only serves to signal the correct check in our `mypy` plugin. So I'm not sure whether it makes sense to use it also in the `forward` and `init` functions. It's more consistent, but a bit weird to read. And for the `backprop`,
```
def backprop(dY: XY_YZ_OutT) -> InT
```
I don't think it makes sense to type `dY` like this? What happens if we just stick to the original `OutT` typing?

Reply:
My issue with the original typing is that it only made sense when you chaining together exactly two layers. `OutT` was the output type of the second layer and was used in `forward()` and `init()` even though it could be wrong for a chained model consisting of more than two layers.

Should not `dY` be the output type of the model, so that if we specify it in the forward method, we should also be specifying it in the backprop?

Reply:
You can't define a `TypeVar` with another generic `TypeVar` as its bound. It would be possible to define `OutT` as a simple alias of `XY_YZ_OutT` though: do you think that would be clearer? And am I right in that you are only talking about the parameter types and return types of these methods and that the type defined within the model passed to them should stay as `XY_YZ_OutT`?

Reply:
Neither solution is 100% ideal, but I personally do think that a separate `OutT` would be clearer, and shorter, making the rest of the code more readible.

Reply:
I've made the change, but am not 100% whether it's exactly what you had in mind.

Reply:
I would basically keep using `OutT` throughout `forward` and `init`, instead of `XY_YZ_OutT`, which I would only use in the constructor to trigger the `mypy` plugin.